### ISSUE TITLE
Zero size bounding box for some TGeoVolumeAssembly nodes after strong alignment

### ISSUE THREAD

**Issue by PALoizeau** [FairRootGroup/FairRoot#1243]
**Describe the bug**
After alignment is applied, the BoundingBox of TGeoVolumeAssembly objects may end up with 0 size (GetDX, GetDY and GetDZ returning 0) if the offset exceed the original boundaries. This is not expected as it means there are other effects on the Volume aside from the translation/rotation.
We (me and @fuhlig1) suspect that this may partially be recursive (affecting the parent or daughter of the aligned volume)

**To Reproduce**
Steps to reproduce the behavior with the attached files (includes example both for problem and for macro-level fix):
1. Compile a copy of FairSoft apr21p2 and FairRoot v18.6.7 (would probably also work with any older versions of each and with recent version of FairRoot)
2. Download [BboxAlignBug.zip](https://github.com/FairRootGroup/FairRoot/files/9830426/BboxAlignBug.zip) and unzip the macro and the two root files in a folder
3. Go to this folder, load your FairSoft-FairRoot environment
4. Run 
    ```
    root -l -b -q 'BboxAlignBug.C("mcbm_beam_2022_05_23_nickel.geo.root", "AlignmentMatrices_mcbm_beam_2022_05_23_nickel.root")'
    ```

**Expected behavior**
After alignment of the geometry, all volumes/nodes are only offset/rotated without changes to their other properties.
This can be cured by forcing the recomputation of the BoundingBox of the affected Volume.

A temp fix for now (until this is hopefully fixed in ROOT) is to simply  do the recompute for all TGeoVolumeAssembly in the geometry tree, which seems to be fast enough that we do not have to try to track affected Volumes to be more selective.

**System information:**
 - OS: Debian 8, Debian 10, Ubuntu 20.04
 - Compiler: GCC > 8
 - Environment: FairSoft apr21 (and patch releases)

**Additional context**
- Corresponding pull request with temp fix on FairRoot side coming soon (waiting until issue done to have proper linking)
- Probably of interest to @TobiasStockmanns and @RalfKliemt  to check that our proposed fix does not break something for PandaRoot

**Comment by PALoizeau** [comment]
I switch the PR back to draft mode for the following reasons:
1. @fuhlig1 has some doubt about the universal loop, as it creates a physical node for all AssemblyVolumes, including those who did not have one to start, which will increase the memory
2. the failing tests for tutorial 4 which is based around geometries and alignment could well be an indication that there is something done there avoiding our problem, and that the proposed fix breaks this solution

=> We will take more time and try to find somebody on our side who could look at this (and understand better than me what is done/happening in these ROOT geo world)

**Comment by PALoizeau** [comment]
1. addressed with new version which is now recomputing the BBox only for VolumeAssembly already having a PhysicalNode.
2. seems to be linked to the standard deviation of pullx being out of the boundary hard-coded at the end of `examples/simulation/Tutorial4/parameters/run_reco.C`.
When looking a the `examples/simulation/Tutorial4/parameters/example.par` file, I have the feeling that the new value of `0.531736` fits better the misalignment introduce than the check against `0.1 +/- 0.03`, but that would be only if the tutorial aims to show how to misalign an ideal geometry. (Openoffice Calc gives me `~0.537931` as standard deviation of the values in `ShiftX` and `~0.553911` for `ShiftY`)
=> If the example purpose is instead to realign a no-ideal geometry, then our fix is breaking the alignment method of the example

**PS:  the Tutorial 4 plot Dx/Dy/Dz outcome is**
- `(0.102392, 0.103786, 0)` after a run of the current dev version
- `(0.531736, 0.556309, 0)` after adding the `gGeoManager->RefreshPhysicalNodes();` line (which does not cure the VolumeAssembly BBox problem)
- `(0.531736, 0.556309, 0)` after adding the call to the VolumeAssembly recomputation on top
- `(0.102392, 0.103786, 0)` again if only the call to the VolumeAssembly recomputation is added

**Comment by fuhlig1** [comment]
@karabowi,

could you please have a look. I think you add the macros to the CI tests.

**Comment by PALoizeau** [comment]
In case it helps with the debugging, here is the patch I used to get more output from plot.C:
[Example4_printout.patch.txt](https://github.com/FairRootGroup/FairRoot/files/9850232/Example4_printout.patch.txt)

**Comment by ChristianTackeGSI** [comment]
> In case it helps with the debugging, here is the patch I used to get more output from plot.C: [Example4_printout.patch.txt](https://github.com/FairRootGroup/FairRoot/files/9850232/Example4_printout.patch.txt)

I have some unrelated cleanups for this `plots.C` in a pipeline. May I include those "prints" there and just reference `@PALoizeau`  in the commit comment? Alternatively, please post a PR for those prints!

I have no idea about the FIXTURES yet… @karabowi needs to look at those. But maybe include those in the mini-PR as well?

**Comment by PALoizeau** [comment]
> > In case it helps with the debugging, here is the patch I used to get more output from plot.C: [Example4_printout.patch.txt](https://github.com/FairRootGroup/FairRoot/files/9850232/Example4_printout.patch.txt)
> 
> I have some unrelated cleanups for this `plots.C` in a pipeline. May I include those "prints" there and just reference `@PALoizeau` in the commit comment? Alternatively, please post a PR for those prints!
> 
> I have no idea about the FIXTURES yet… @karabowi needs to look at those. But maybe include those in the mini-PR as well?

Sure you can include the printouts in your own general review of the macro, close to half of my commits to Fairroot are anyway originally ideas from Florian that he had no time to implement/test (^^)'

For the `FIXTURES` it is something we use in CBM that I copied here because it I did not carefully read the example and always forgot one of the previous steps. 
From what we learned adding them in CBM with the goal of making our CI tests also "examples of how to run", it is probably better to do this in a wider PR improving all tests, for example protecting all tests chains and adding also a cleanup of the output file(s) step. 
Quite often in the CI context we do not keep the build folder and it does not matter what we leave behind, but new users tend to re-run the tests multiple time after small changes and get confused by "updating" parameter and/or output files.

**Comment by karabowi** [comment]
@PALoizeau , if you feel that the new standard deviation fits better the expectation, then change the plots.C macro to fit the expected value, f.e. to 0.5+-0.1.

**Comment by PALoizeau** [comment]
> @PALoizeau , if you feel that the new standard deviation fits better the expectation, then change the plots.C macro to fit the expected value, f.e. to 0.5+-0.1.

@fuhlig1 Is this still intended to be included in the next FairRoot release or is the priority lower now that our CbmRoot side fix seems to work (but is not really tested due to larger uncertainties in the mCBM geometries/data)?

In the first case I should probably plan to the following in the coming days?
- recheck the "tests success" values more carefully in the input plots and in the ones after mis-alignement+re-alignement, 
- confirm that our fix does not break the expected behavior
- update the test values where needed and try to get the PR green

**Comment by karabowi** [comment]
Issue fixed by merging the PR #1244.
